src/Translator/Extractor/AdminExtractor.php

```diff
@@ -0,0 +1,158 @@
+
+use Psr\Log\LoggerInterface;
+use Sonata\AdminBundle\Admin\BreadcrumbsBuilderInterface;
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Translator\LabelTranslatorStrategyInterface;
+use Symfony\Component\Translation\Extractor\ExtractorInterface;
+use Symfony\Component\Translation\MessageCatalogue;
+
+/**
+ * @internal
+ */
+final class AdminExtractor implements ExtractorInterface, LabelTranslatorStrategyInterface
+{
+    /**
+     * @var string
+     */
+    private $prefix;
+
+    /**
+     * @var MessageCatalogue|null
+     */
+    private $catalogue;
+
+    /**
+     * @var LoggerInterface|null
+     */
+    private $logger;
+
+    /**
+     * @var Pool
+     */
+    private $adminPool;
+
+    /**
+     * @var LabelTranslatorStrategyInterface|null
+     */
+    private $labelStrategy;
+
+    /**
+     * @var string|null
+     */
+    private $domain;
+
+    /**
+     * @var BreadcrumbsBuilderInterface
+     */
+    private $breadcrumbsBuilder;
+
+    public function __construct(Pool $adminPool, BreadcrumbsBuilderInterface $breadcrumbsBuilder, LoggerInterface $logger = null)
+    {
+        $this->logger = $logger;
+        $this->adminPool = $adminPool;
+        $this->breadcrumbsBuilder = $breadcrumbsBuilder;
+    }
+
+    public function extract($resource, MessageCatalogue $catalogue)
+    {
+        $this->catalogue = $catalogue;
+
+        foreach ($this->adminPool->getAdminGroups() as $name => $group) {
+            $catalogue->set($name, $this->prefix.$name, $group['label_catalogue']);
+        }
+
+        foreach ($this->adminPool->getAdminServiceIds() as $id) {
+            $admin = $this->adminPool->getInstance($id);
+
+            $this->labelStrategy = $admin->getLabelTranslatorStrategy();
+            $this->domain = $admin->getTranslationDomain();
+
+            $label = $admin->getLabel();
+            if (!empty($label)) {
+                $catalogue->set($label, $this->prefix.$label, $admin->getTranslationDomain());
+            }
+
+            $admin->setLabelTranslatorStrategy($this);
+
+            // call the different public method
+            $methods = [
+                'getShow',
+                'getDatagrid',
+                'getList',
+                'getForm',
+            ];
+
+            $actions = [
+                'list',
+                'edit',
+                'create',
+                'update',
+                'batch',
+                'delete',
+            ];
+
+            if ($this->logger) {
+                $this->logger->info(sprintf('Retrieving message from admin:%s - class: %s', $admin->getCode(), \get_class($admin)));
+            }
+
+            foreach ($methods as $method) {
+                try {
+                    $admin->$method();
+                } catch (\Exception $e) {
+                    if ($this->logger) {
+                        $this->logger->error(sprintf('ERROR : admin:%s - Raise an exception : %s', $admin->getCode(), $e->getMessage()));
```

Comment:
Please prefer using the context array rather than making messages dynamic, and using the magical `exception` key.
```suggestion
                        $this->logger->error('Error while trying to call method on admin', [
                            'admin' => $admin,
                            'exception' => $e,
                        ]);
```